subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/AbstractVirtualFileSystem.java

```diff
@@ -45,31 +49,58 @@ public Optional<MetadataSnapshot> findMetadata(String absolutePath) {
     }
 
     @Override
-    public void store(String absolutePath, FileSystemLocationSnapshot snapshot) {
-        rootReference.update(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+    public FileSystemLocationSnapshot store(String absolutePath, Supplier<FileSystemLocationSnapshot> snapshotSupplier) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(absolutePath);
+        FileSystemLocationSnapshot snapshot = snapshotSupplier.get();
+        updateCheckingVersion(absolutePath, versionBefore, snapshot);
+        return snapshot;
+    }
+
+    @Override
+    public FileSystemLocationSnapshot store(String baseLocation, StoringAction storingAction) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(baseLocation);
+        return storingAction.snapshot(snapshot -> updateCheckingVersion(snapshot.getAbsolutePath(), versionBefore, snapshot));
+    }
+
+    private void updateCheckingVersion(String absolutePath, long versionBefore, FileSystemLocationSnapshot snapshot) {
+        long versionAfter = versionHierarchyRoot.getVersionFor(absolutePath);
+        // Only update VFS if no changes happened in between
+        if (versionBefore == versionAfter) {
+            rootReference.update(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+        } else {
+            LOGGER.debug("Changes to the virtual file system happened while snapshotting '{}', not storing resulting snapshot", absolutePath);
+        }
     }
 
     @Override
     public void invalidate(Iterable<String> locations) {
         LOGGER.debug("Invalidating VFS paths: {}", locations);
         rootReference.update(root -> {
```

Comment:
This is somewhat misleading. Looking at this code I expected this to be calling `AtomicReference.update()` which requires a side-effect-free update function, as it gets re-executed as needed. What is happening here is more of an `updateUnderLock()`.

We could also just inline this type. We don't really need to pass this around, and instead `AbstractVirtualFileSystem` could create the instance in its constructor based on the `caseSensitivity`.

I think doing so would significantly reduce code complexity, and would allow us to remove this somewhat weird type altogether. It would even make the constructor of `AbstractVirtualFileSystem` more symmetric.

Reply:
I think it makes sense to inline the lock and to guard the whole state of the VFS with this one lock. That is what we have been doing anyway with this method. Let's do that in a follow up PR.